tests/virt_autotest/xen_guest_irqbalance.pm

```diff
@@ -0,0 +1,159 @@
+# SUSE's openQA tests
+
+# Copyright 2022 SUSE LLC
+# SPDX-License-Identifier: GPL-2.0-or-later
+#
+# Summary: xen domU irqbalance test
+# Maintainer: Julie CAO <JCao@suse.com>
+
+use base "virt_feature_test_base";
+use strict;
+use warnings;
+use testapi;
+use version_utils qw(is_sle);
+use set_config_as_glue;
+use virt_autotest::common;
+use virt_autotest::utils qw(is_kvm_host guest_is_sle wait_guest_online print_cmd_output_to_file);
+
+sub run_test {
+    my $self = shift;
+
+    return if is_kvm_host;
+    #save original guest configuration file in case of restore in post_fail_hook()
+    my $downloaded_xml_dir = "/tmp/download_vm_xml";
+    foreach my $guest (keys %virt_autotest::common::guests) {
+        unless (script_run("ls $downloaded_xml_dir/$guest.xml") == 0) {
+            assert_script_run "mkdir -p $downloaded_xml_dir";
+            assert_script_run "virsh dumpxml --inactive $guest > $downloaded_xml_dir/$guest.xml";
+        }
+    }
+
+    foreach my $guest (keys %virt_autotest::common::guests) {
+
+        #irqbalance test run on sle12sp3+ guest
+        return if guest_is_sle($guest, '<=12-sp2');
+
+        record_info("Test $guest");
+        prepare_guest_for_irqbalance($guest);
+
+        my $nproc = script_output("ssh root\@$guest 'nproc'");
+
+        #Check if network interrupts are distributed over CPUs in guest initally as irqbalance is enabled by default
+        my @initial_total_irqs_on_cpu = get_nic_irqs_distribution_from_guest($guest, $nproc);
+        for (my $cpu_id = 0; $cpu_id < $nproc; $cpu_id++) {
+            die("The value of network interrupts for CPU" . $cpu_id . " is 0") if ($initial_total_irqs_on_cpu[$cpu_id] == 0);
```

Comment:
On extreme conditions this won't be true, eg vcpu numbers is big while interrupts are lower than vcpu number. Suggest to skip this check.

Reply:
If we skip this check, we will missing the check if irqbalance works by default. because after NIC pings, ''irqbalance --oneshot' is excuted in guests. one feasible way is to generate more NIC irqs before we do this check, what do you think?

Reply:
IMHO, the latter designed test flow can verify if irqbalance works when the service is enabled. Here if we want to have "default" tested, maybe checking if the service is running is enough? BTW, this script is better to be designed open/workable for any guest that comes to be tested, so it can face many different guest configuration&situations, possible to even have the service disabled by autoyast during guest installation.

Reply:
no, the check after NIC binding is to check the irqbalance tool, 'irqbalance --oneshot'. the tool and service may not be totally equal, but the function should be equal. 

irqbalance.service is set to running before testing is run in my test in sub prepare_guest_for_irqbalance(). so all guests will be ok for irqbalance test.
```
assert_script_run "ssh root\@$vm_name \"zypper in -y irqbalance\"" unless script_run("ssh root\@$vm_name \"rpm -q irqbalance\"") eq 0;
assert_script_run "ssh root\@$vm_name \"systemctl start irqbalance.service\"" unless script_run("ssh root\@$vm_name \"systemctl status irqbalance\"") eq 0;
```